### PR TITLE
docs: improve template documentation

### PR DESCRIPTION
**Pull request by jonaslagoni**
**Description**
This PR adds types to all arguments to improve the template developer experience and add contribution guidelines.

Blocked by #69

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Overview
+To contribute to this project, follow the rules from the general [CONTRIBUTING.md](https://github.com/asyncapi/.github/blob/master/CONTRIBUTING.md) document in the community repository.
+
+## Getting started
+Extending the contribution guidelines above here is how to quick start your contribution process.
+* The template is rendered using the AsyncAPI generator React renderer, [the authoring docs](https://github.com/asyncapi/generator/blob/master/docs/authoring.md#react-1) is great way to quick start your template journey to understand the setup.
+* The template support two examples generations which is used to inspect your changes. These can be generated using the scripts `npm run generate:examples`.
+* Do NOT commit the examples on your branch. Once your PR are merged the examples will be auto generated and committed automatically by the CI. This is to ease the PR review process.
```
Comment (magicmatatjahu) [inline CONTRIBUTING.md:8]: Sometimes generating examples will be great if you wanna show changes to reviewers. I think it should be included here with annotation that before merge it should be removed.

File: components/channel/ChannelParameterUnwrap.js [hunk 2]
```diff
@@ -18,8 +20,8 @@ import { camelCase, castToTsType, realizeChannelNameWithoutParameters } from '..
   const streetlightIdParam = "" + channel.substring(0, streetlightIdEnd);
  * 
  * 
- * @param {*} channelName to be unwrapped
- * @param {*} channelParameters the parameters which are to be unwrapped from the NATS topic.
+ * @param {string} channelName to be unwrapped
+ * @param {{[key: string]: ChannelParameter}} channelParameters the parameters which are to be unwrapped from the NATS topic.
```
Comment (magicmatatjahu) [inline components/channel/ChannelParameterUnwrap.js:24]:
```suggestion
 * @param {Object.<string, ChannelParameter>} channelParameters the parameters which are to be unwrapped from the NATS topic.
```

```
Object.<key, value>
```

is a proper way to define dictionary in JSDoc

File: components/channel/publish.js [hunk 3]
```diff
@@ -1,13 +1,15 @@
...
  * Component which returns a function which publishes to the given channel
  * 
- * @param {*} defaultContentType 
- * @param {*} channelName to publish to
- * @param {*} message which is being published
- * @param {*} channelParameters parameters to the channel
+ * @param {string} defaultContentType 
+ * @param {string} channelName to publish to
+ * @param {Message} message which is being published
+ * @param {{[key: string]: ChannelParameter}} channelParameters parameters to the channel
```
Comment (magicmatatjahu) [inline components/channel/publish.js:12]:
```suggestion
 * @param {Object.<string, ChannelParameter>} channelParameters parameters to the channel
```

File: components/channel/reply.js [hunk 4]
```diff
@@ -2,16 +2,25 @@ import { OnSendingData } from './OnSendingData';
...
- * @param {*} channelName to reply to
- * @param {*} replyMessage which is being send as a reply
- * @param {*} receiveMessage which is being received
- * @param {*} channelParameters parameters to the channel
- * @param {*} params template parameters
+ * @param {string} defaultContentType 
+ * @param {string} channelName to reply to
+ * @param {Message} replyMessage used to reply to request
+ * @param {Message} receiveMessage which is received by the request 
+ * @param {{[key: string]: ChannelParameter}} channelParameters parameters to the channel
```
Comment (magicmatatjahu) [inline components/channel/reply.js:22]:
```suggestion
 * @param {Object.<string, ChannelParameter>} channelParameters parameters to the channel
```

File: components/channel/request.js [hunk 5]
```diff
@@ -1,17 +1,19 @@
...
- * @param {*} defaultContentType 
- * @param {*} channelName to request to
- * @param {*} requestMessage which should be send
- * @param {*} receiveMessage which is received after request
- * @param {*} channelParameters parameters to the channel
+ * @param {string} defaultContentType 
+ * @param {string} channelName to request to
+ * @param {Message} requestMessage used to send the request
+ * @param {Message} replyMessage which is receive in the reply
+ * @param {{[key: string]: ChannelParameter}} channelParameters parameters to the channel
```
Comment (magicmatatjahu) [inline components/channel/request.js:14]:
```suggestion
 * @param {Object.<string, ChannelParameter>} channelParameters parameters to the channel
```

File: components/channel/subscribe.js [hunk 6]
```diff
@@ -1,17 +1,18 @@
...
  * Component which returns a function which subscribes to the given channel
  * 
- * @param {*} defaultContentType 
- * @param {*} channelName to subscribe to
- * @param {*} message which is being received
- * @param {*} channelParameters parameters to the channel
+ * @param {string} defaultContentType 
+ * @param {string} channelName to subscribe to
+ * @param {Message} message which is being received
+ * @param {{[key: string]: ChannelParameter}} channelParameters parameters to the channel
```
Comment (magicmatatjahu) [inline components/channel/subscribe.js:13]:
```suggestion
 * @param {Object.<string, ChannelParameter>} channelParameters parameters to the channel
```

File: components/index/publish.js [hunk 7]
```diff
@@ -1,13 +1,15 @@
...
- * @param {*} defaultContentType 
- * @param {*} channelName to publish to
- * @param {*} message to publish
- * @param {*} messageDescription 
- * @param {*} channelParameters parameters to the channel
+ * @param {string} defaultContentType 
+ * @param {string} channelName to publish to
+ * @param {Message} message which is being published
+ * @param {string} messageDescription 
+ * @param {{[key: string]: ChannelParameter}} channelParameters parameters to the channel
```
Comment (magicmatatjahu) [inline components/index/publish.js:12]:
```suggestion
 * @param {Object.<string, ChannelParameter>} channelParameters parameters to the channel
```

File: components/index/reply.js [hunk 8]
```diff
@@ -1,14 +1,24 @@
...
- * @param {*} receiveMessage which is to be returned 
- * @param {*} messageDescription 
- * @param {*} channelParameters parameters to the channel
- * @param {*} params passed template parameters 
+ * @param {string} defaultContentType 
+ * @param {string} channelName to setup reply to
+ * @param {Message} replyMessage used to reply to request
+ * @param {Message} receiveMessage which is received by the request 
+ * @param {string} messageDescription 
+ * @param {{[key: string]: ChannelParameter}} channelParameters parameters to the channel
```
Comment (magicmatatjahu) [inline components/index/reply.js:20]:
```suggestion
 * @param {Object.<string, ChannelParameter>} channelParameters parameters to the channel
```

File: components/index/request.js [hunk 9]
```diff
@@ -1,14 +1,16 @@
...
- * @param {*} requestMessage to request
- * @param {*} replyMessage to receive
- * @param {*} messageDescription 
- * @param {*} channelParameters parameters to the channel
+ * @param {string} defaultContentType 
+ * @param {string} channelName to request to 
+ * @param {Message} requestMessage used to send the request
+ * @param {Message} replyMessage which is receive in the reply
+ * @param {string} messageDescription 
+ * @param {{[key: string]: ChannelParameter}} channelParameters parameters to the channel
```
Comment (magicmatatjahu) [inline components/index/request.js:13]:
```suggestion
 * @param {Object.<string, ChannelParameter>} channelParameters parameters to the channel
```

File: components/index/subscribe.js [hunk 10]
```diff
@@ -1,13 +1,15 @@
...
- * @param {*} defaultContentType 
- * @param {*} channelName to publish to
- * @param {*} message received
- * @param {*} messageDescription 
- * @param {*} channelParameters parameters to the channel 
+ * @param {string} defaultContentType 
+ * @param {string} channelName to publish to
+ * @param {Message} message which is being received
+ * @param {string} messageDescription 
+ * @param {{[key: string]: ChannelParameter}} channelParameters parameters to the channel
```
Comment (magicmatatjahu) [inline components/index/subscribe.js:12]:
```suggestion
 * @param {Object.<string, ChannelParameter>} channelParameters parameters to the channel
```

**Review by magicmatatjahu — CHANGES_REQUESTED** [review]
Very good! But please also change all occurrences of `{[key: string]: ChannelParameter}` to the `{Object.<string, ChannelParameter>` in JSDOc. There are more of them than in my suggestions.

File: CONTRIBUTING.md (hunk 1 above)
Comment (jonaslagoni) [reply CONTRIBUTING.md:8]: good point 🤔

Comment (jonaslagoni) [reply CONTRIBUTING.md:8]: @magicmatatjahu something like this?
```
* In most cases committing changes to the examples is a good thing, but try not to clutter up the PR. You may be asked to remove committed example code to remove clutter and ease the review process. Once your PR is merged the examples will be auto-generated and committed automatically by the CI. This is to ease the PR review process.
```

File: components/channel/ChannelParameterUnwrap.js (hunk 2 above)
Comment (jonaslagoni) [reply components/channel/ChannelParameterUnwrap.js:24]: Did not know that, thanks! I'll do a search/replace 😄

File: CONTRIBUTING.md (hunk 1 above)
Comment (magicmatatjahu) [reply CONTRIBUTING.md]: @jonaslagoni Yes, it is good :) But add note that generated examples should be removed before merge - reviewers can miss about that.

Comment (jonaslagoni) [reply CONTRIBUTING.md]: @magicmatatjahu done :)

**Review by magicmatatjahu — APPROVED** [review]
LGTM

**Comment by jonaslagoni** [comment]
Thanks @magicmatatjahu 🙇

**Comment by jonaslagoni** [comment]
@all-contributors please add @jonaslagoni for bug, maintenance, code, doc, ideas, question, test

**Comment by jonaslagoni** [comment]
@all-contributors please add @magicmatatjahu for review

**Comment by jonaslagoni** [comment]
@all-contributors please add @magicmatatjahu for review

**Pull request merged**